[DOC-11986] Rework the Student Record Tutorial for the Server docs (#167)
couchbaselabs/docs-devex: modules/tutorials/pages/java-tutorial/retrieve-records.adoc

Reviewer: Again, I would rename this back to `retrieving-documents.adoc` and remove the page alias

+
[source, json]
----
[
  {
    "course-name": "art history",
    "credit-points": 100,
    "faculty": "fine art",
    "id": "ART-HISTORY-000001"
  },
  {
    "course-name": "fine art",
    "credit-points": 50,
    "faculty": "fine art",
    "id": "FINE-ART-000002"
  }
]
----

[#retrieve-with-sdk]
== With the SDK

You can also use {sqlpp} queries to retrieve your records with the SDK.
Unlike the query editor, though, you must include the name of the bucket and the scope to fully qualify the name of the collection in the {sqlpp} statement in your application. 
For example:

[source, sqlpp]
----
select crc.* from `student-bucket`.`art-school-scope`.`course-record-collection` crc
----

=== Retrieve Your Records

You can use the SDK to retrieve all course records at once, or narrow your search down to retrieve records based on specific criteria.

==== Retrieve All Course Records

To retrieve all of your course records using the Java SDK:

. In your `java` directory, create a new file called `ArtSchoolRetriever.java`.
. Paste the following code block into your `ArtSchoolRetriever.java` file:
+
[source, java]
----
include::example$ArtSchoolRetrieverAll.java[]
----

> +

Reviewer: 1. Add a step here telling users to change line 10 of the file to match the user name and password they chose when they set up the cluster, or
2. Change the instructions in `tutorial-install-server.adoc` telling the users explicitly which user name and password they should use when setting up this cluster

... or both!

. Open a terminal window and navigate to your `student` directory.
. Run the command `mvn install` to pull in all the dependencies and rebuild your application.
. Run the following command to retrieve all course records:
+
[source, sh]
----
mvn exec:java -Dexec.mainClass="ArtSchoolRetriever" -Dexec.cleanupDaemonThreads=false
----
+
If the retrieval is successful, the course information outputs in the console log.
+
image::record-retrieval-console-output.png[alt="Console showing successful course retrieval using the SDK"]

==== Retrieve Course Records with Less than 200 Credits

You can set parameters in your code to narrow your search down further.
To retrieve only courses with less than 200 `credit-points` using the Java SDK:

. In your `java` directory, create a new file called `ArtSchoolRetriever.java`.

> . Paste the following code block into your `ArtSchoolRetriever.java` file:

Reviewer: Needs new name, we've already created this file:

```suggestion
. In your `java` directory, create a new file called `ArtSchoolRetrieverParameters.java`.
. Paste the following code block into your `ArtSchoolRetrieverParameters.java` file:
```

+
[source, java]
----
include::example$ArtSchoolRetriever.java[]
----

> +

Reviewer: 1. Add a step here telling users to change line 11 of the file to match the user name and password they chose when they set up the cluster, or
2. Change the instructions in `tutorial-install-server.adoc` telling the users explicitly which user name and password they should use when setting up this cluster

... or both!

. Open a terminal window and navigate to your `student` directory.
. Run the command `mvn install` to pull in all the dependencies and rebuild your application.
. Run the following command to retrieve all course records:
+
[source, sh]
----

> mvn exec:java -Dexec.mainClass="ArtSchoolRetrieverParameters" -Dexec.cleanupDaemonThreads=false

Reviewer: I'm consistently getting timeouts with this command. Might need to talk to @RichardSmedley for suggestions here.

Author: That's strange... I'm getting a successful build every time I run it. Maybe Richard has an idea of what could be going on.

= Retrieve Records
:description: Retrieve records, or documents, from your collections using {sqlpp}, Couchbase's SQL-based query language.
:page-topic-type: tutorial
:page-pagination: full
:page-toclevels: 2
:imagesdir: ../../images
:page-aliases: retrieving-documents.adoc

[abstract]
{description}
You can retrieve your records using the <<#retrieve-with-query-editor,query editor>> or the <<#retrieve-with-sdk,SDK>>.

[#retrieve-with-query-editor]
== With the Query Editor

> 

Reviewer 2: I think Vale should have flagged the stacked headings - can you think of even a single sentence to add here so we don't have two headings with nothing in between?

Reviewer 2: Maybe an overview of the substeps, like: 

To retrieve the records with the Query Editor, you must first define an index.

=== Define an Index

Before you can retrieve your records with the query editor, you must first define an index in your bucket.
The index helps your database find specific data when you run a query.

To define an index:

. In your Couchbase database in your browser, click menu:Query[] to open the query editor.
. Select *student-bucket* and *art-school-scope* in the *context* drop-down.

> Using these filters means you can narrow down the scope of your queries, and that you don't need to add the names of your bucket and scope to your queries.

Reviewer 2: ```suggestion
Using these filters means you can narrow down the scope of your queries. 
You do not need to add the names of your bucket and scope to your queries.
```

+
image::query-editor-filters.png[alt="Filters in the Query Editor"]
+
. Enter the following query into your query editor:
+
[source, sqlpp]
----
create primary index course_idx on `course-record-collection`
----
+
. Click btn:[Execute] to create a single index called `course_idx` in your `course-record-collection`.

=== Retrieve Your Records

> 

Reviewer 2: Same note - can we add even a single sentence here?

+
[source, json]
----
[
  {
    "course-name": "art history",
    "credit-points": 100,
    "faculty": "fine art"
  },
  {
    "course-name": "fine art",
    "credit-points": 50,
    "faculty": "fine art"
  }
]
----

=== Retrieve Record IDs

The `id` field is not automatically returned when you retrieve all of your course information.

The `id` is part of a document's meta structure, and to retrieve it you must adjust your {sqlpp} query and run it again:

. Enter the following query into your query editor:
+
[source, sqlpp]
----
select META().id, crc.* from `course-record-collection` crc where crc.`credit-points` < 200
----
+
The `META()` function call returns any property contained inside the document's metadata, including the ID.
+
. Click btn:[Execute] to retrieve course records and their IDs.
+
[source, json]
----
[
  {
    "course-name": "art history",
    "credit-points": 100,
    "faculty": "fine art",
    "id": "ART-HISTORY-000001"
  },
  {
    "course-name": "fine art",
    "credit-points": 50,
    "faculty": "fine art",
    "id": "FINE-ART-000002"
  }
]
----

[#retrieve-with-sdk]
== With the SDK

You can also use {sqlpp} queries to retrieve your records with the SDK.

> Unlike the query editor, though, you must include the name of the bucket and the scope to fully qualify the name of the collection in the {sqlpp} statement in your application. For example:

Reviewer 2: ```suggestion
Unlike the query editor, though, you must include the name of the bucket and the scope to fully qualify the name of the collection in the {sqlpp} statement in your application. 
For example:
```

[source, sqlpp]
----
select crc.* from `student-bucket`.`art-school-scope`.`course-record-collection` crc
----

=== Retrieve Your Records

> 

Reviewer 2: Same thing - some content in between headings, please!

==== Retrieve All Course Records

To retrieve all of your course records using the Java SDK:

. In your `java` directory, create a new file called `ArtSchoolRetriever.java`.
. Paste the following code block into your `ArtSchoolRetriever.java` file:
+
[source, java]

> ----

Reviewer 2: Same thing about sample code. For longer samples, it's better to pull in via include - for the single line examples, less important (but still something to consider)

            for (JsonObject row : result.rowsAsObject()) {
                System.out.println("Found row: " + row);
            }

        } catch (CouchbaseException ex) {
            ex.printStackTrace();
        }
    }
}
----
+
. Open a terminal window and navigate to your `student` directory.
. Run the command `mvn install` to pull in all the dependencies and rebuild your application.
. Run the following command to retrieve all course records:
+
[source, sh]
----
mvn exec:java -Dexec.mainClass="ArtSchoolRetriever" -Dexec.cleanupDaemonThreads=false
----
+
If the retrieval is successful, the course information outputs in the console log.
+
image::record-retrieval-console-output.png[alt="Console showing successful course retrieval using the SDK"]

==== Retrieve Course Records with Less than 200 Credits

You can set parameters in your code to narrow your search down further.
To retrieve only courses with less than 200 `credit-points` using the Java SDK:

. In your `java` directory, create a new file called `ArtSchoolRetriever.java`.
. Paste the following code block into your `ArtSchoolRetriever.java` file:
+
[source, java]

> ----

Reviewer 2: Code again.

import com.couchbase.client.java.query.QueryResult;

public class ArtSchoolRetrieverParameters {

    public static void main(String[] args) {
        Cluster cluster = Cluster.connect("localhost",
                "Administrator", "password");

        retrieveCourses(cluster);

        cluster.disconnect();
    }

    private static void retrieveCourses(Cluster cluster) {

        try {
            final QueryResult result = cluster.query("select crc.* " +
                            "from `student-bucket`.`art-school-scope`.`course-record-collection` crc " +
                            "where crc.`credit-points` < $creditPoints", // <1>   
                    QueryOptions.queryOptions()
                            .parameters(JsonObject.create().put("creditPoints", 200))); // <2>    

            for (JsonObject row : result.rowsAsObject()) {
                System.out.println("Found row: " + row);
            }

        } catch (CouchbaseException ex) {
            ex.printStackTrace();
        }
    }
}
----
+
<1> This {sqlpp} statement takes the parameter `$creditPopints`, which is then substituted by the value in the second parameter when the statement is called.

> <2> The second parameter in the function call, with a value that replaces `$creditPoints`.

Reviewer 2: Same thing on code callouts.

+
. Open a terminal window and navigate to your `student` directory.
. Run the command `mvn install` to pull in all the dependencies and rebuild your application.
. Run the following command to retrieve all course records:
+
[source, sh]
----
mvn exec:java -Dexec.mainClass="ArtSchoolRetrieverParameters" -Dexec.cleanupDaemonThreads=false
----
+
If the retrieval is successful, the course information with your parameters outputs in the console log.
+
image::record-retrieval-parameters-console-output.png[alt="Console showing successful course retrieval using parameters using the SDK"]

> If you come across errors in your console, check out the xref:java-tutorial/tutorial-troubleshooting.adoc[troubleshooting page].

Reviewer 2: ```suggestion
If you come across errors in your console, see the xref:java-tutorial/tutorial-troubleshooting.adoc[troubleshooting page].
```
